### DISCUSSION TITLE
add ON DELETE CASCADE to routes during delete of service

### DISCUSSION THREAD

**Discussion by darrenjennings** [Kong/kong#7314] [Ideas and feature requests]
### Summary

Suggestion to cascade delete routes when deleting a service. If a service has a lot of routes, it is painful to get `/services/:service/routes` and then `DELETE /routes/:id`. Was not able to see any historical significance of not having cascade delete of routes when deleting a service, and it seems that sometimes we cascade delete other entities. For example, if you create a plugin on a service, then deleting the service also deletes the plugin. @thibaultcha mentioned that we might want to allow for dangling routes while deleting the service so that the router never produces invalid responses. Could this be circumvented by disassociating the service from the route before performing the service delete?

Added migration locally and tested to success:
```sql
ALTER TABLE routes DROP CONSTRAINT routes_service_id_fkey,
        ADD CONSTRAINT routes_service_id_fkey FOREIGN KEY (service_id, ws_id)
        REFERENCES services(id, ws_id) ON DELETE CASCADE;
```

```lua
{ service = { type = "foreign", reference = "services", on_delete = "cascade" }, },
```

It is nice to have safety guards during `DELETE` of services since you could be performing a very destructive action, so curious if it is better to "opt-in" by not creating the DB association but handling at the ORM level e.g. `DELETE /services/:service/routes?cascade=true`. This is setup in other ORM's like e.g. doctrine's cascade="remove" and [orphanRemoval](https://www.doctrine-project.org/projects/doctrine-orm/en/2.7/reference/working-with-associations.html#orphan-removal) so that you would opt-in to the relationship deletion.

**Comment by bungle** [comment]
We have `on_delete=cascade` in some entities:
https://github.com/Kong/kong/blob/28e4baf323a3b01ede1be4301fd490408a1e5d6c/kong/db/schema/entities/plugins.lua#L18-L20

**Comment by bungle** [comment]
The one thing about cascades that worries me is the cache. I think we might have some problems with it already. This same problem is probably also with caching entities that have ttl for longer than ttl. As deletions happen on DB level, there is no `worker_events` or `cluster_events` to be processed, and that can lead to issues, like one node having the entity still somewhere in cache while new node does not have it, which leads to inconsistent behavior between nodes.
